include/ghex/transport_layer/ucx/request.hpp

```diff
@@ -269,7 +269,7 @@ namespace gridtools{
                         return m_req->m_worker->m_parallel_context->thread_primitives().critical(
 			    [this]()
 			    {
-				if if (!(*m_req->m_completed))
+				if if (!(*m_completed))
```

Comment:
@boeschf This is more complicated. We do check before, but between we check and the critical section the request might get completed. Looks like this is not thread safe. We can't access `m_req` without checking `null`, but I think we then need to run a critical section without accessing `m_req`.

Reply:
So the entire `cancel` body should be inside a critical section. This is not a problem for the future-based request, because we only free the request inside the `test/cance` functions.